Fix #9169 refine idle timeout and failure

gregw opened the pull request:
Fix #9169 Only fail request callback if a failure has not been otherwise notified. Slight optimisation for failing idle timeouts by avoiding double lock.

gregw committed:
Fix #9169 refine idle timeout and failure

Only fail request callback if a failure has not been otherwise notified.
Slight optimization for failing idle timeouts by avoiding double lock.

gregw committed:
Fix #9169 refine idle timeout and failure

Only fail request callback if a failure has not been otherwise notified.
Slight optimization for failing idle timeouts by avoiding double lock.

lorban reviewed (approved):
LGTM

lachlan-roberts commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -382,6 +382,10 @@ public Runnable onIdleTimeout(TimeoutException t)
+
+                // If there is no idle nor failure listener, then we can fail the callback directly without double lock
+                if (onIdleTimeout == null && _onFailure == null && _request != null)
+                    return () -> _request._callback.failed(t);
```
Just noting that there is some things `onFailure` would do that aren't done in this optimisation case. 

- making a temp request for logging, and producing an error response if one hasn't been created.
- setting _failure content chunk.

gregw replied:
> Just noting that there is some things `onFailure` would do that aren't done in this optimisation case.
> 
> * making a temp request for logging, and producing an error response if one hasn't been created.
That is only done if the _request is null, hence the not null check at 387

> * setting _failure content chunk.

That is only needed if something is to read content, but there has been no notification to the app that there is something to read and the callback is failed anyway.

So I think this is fine.... but I might set the _failure chunk as well, just in case.

gregw replied:
@lachlan-roberts I refined it a bit so that we always set `_failure` if we are going to fail the callback.

joakime commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -476,7 +480,7 @@ else if (ExceptionUtil.areNotAssociated(_failure.getFailure(), x) && _failure.ge
 
                     // If the application has not been otherwise informed of the failure
-                    if (invokeOnContentAvailable == null && invokeWriteFailure == null)
+                    if (invokeOnContentAvailable == null && invokeWriteFailure == null && onFailure == null)
```
This feels fishy.
Been trying to remind myself of the changes we made in #9684 and when we removed many of these onFailure tests.  Not having much luck.

gregw replied:
@joakime to me it was stranger that that we would notify listeners AND fail the callback.  Such double actions are frequently wrong.    This feels much better to explain: if we have told the app about the failure, then it is still the apps responsibility to complete the callback, but if we have not told them about it then it is ours.

gregw replied:
@joakime I refined the logic and comments a little bit. Less fishy?

joakime replied:
I still have this feeling like we are restoring the behavior from before #9684, but I honestly cannot prove it.
Oh well, all of the tests pass, and we added a bunch more during the #9684 effort, so I'm happy with this as-is.

gregw replied:
> I still have this feeling like we are restoring the behavior from before #9684

I see it as continuing the work of #9684 (and other recent error changes).
Previously we were doing everything on failure: fail any IO operations, called the listeners & failed the callback. 

We have moved to a mechanism where we almost always only do a single notification... except prior to this PR we could call the error callback AND fail the callback.

So this PR really just competes previous work that was reducing dual notification.

We now only get dual notification if you have both demand and a write outstanding.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/12/9169/idletimeout

gregw committed:
Fix #9169 refine idle timeout and failure

Only fail request callback if a failure has not been otherwise notified.
Slight optimization for failing idle timeouts by avoiding double lock.
Always create a failure if failing the callback.

gregw committed:
Fix #9169 refine idle timeout and failure

Only fail request callback if a failure has not been otherwise notified.
Slight optimization for failing idle timeouts by avoiding double lock.
Always create a failure if failing the callback.

gregw commented:
@sbordet please have a look at this first thing, as I think it is a contender to slip into 12.0.1

gregw committed:
Merge branch 'jetty-12.0.x' into fix/12/9169/idletimeout

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -380,29 +379,29 @@ public Runnable onIdleTimeout(TimeoutException t)
+                    {
+                        if (onIdleTimeout.test(t))
+                        {
+                            // If the idle timeout listener(s) return true, then we call onFailure and any task it returns.
```
... and **run** any task...

sbordet reviewed (approved):
Just a nit.

gregw committed:
Fix #9169 refine idle timeout and failure

Only fail request callback if a failure has not been otherwise notified.
Slight optimization for failing idle timeouts by avoiding double lock.
Always create a failure if failing the callback.
